### PR TITLE
Enable plugin approach to enable default experiments framework

### PR DESCRIPTION
**Pull request by sunishsheth2009**
## What changes are proposed in this pull request?
Enable plugin approach to enable default experiments framework

## How is this patch tested?
- [x] Created a job with the mlflow wheel: and [creating the job experiment was successful](https://dbc-98e50cb7-ce96.dev.databricks.com/?o=868247350988813#job/429899706490992/run/261100988457345).
- [x] Created a [notebook experiment](https://dbc-98e50cb7-ce96.dev.databricks.com/?o=868247350988813#notebook/1142404851242222/command/1142404851242223) as well.
- [x] Tests added

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/_default_experiment/registry.py
Comment (sunishsheth2009) [inline mlflow/tracking/_default_experiment/registry.py:66]: Will remove this TODO in later PR, tried removing it here but that caused many tests failures. So didn't want to complicate this PR

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:13]: `_get_experiment_id()` is also used in other places, e.g. MLflow Projects. Can we discuss how it's used by `_get_experiment_id()` and cite ``mlflow.start_run()`` as just one example of where `_get_experiment_id()` is used?

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:7]:
```suggestion
    Abstract base class for objects that provide the ID of an MLflow experiment based on the
    current client environment. For example, when the MLflow client is running in a Databricks
    Job, a provider is used to obtain the ID of the MLflow experiment associated with the Job.
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:23]:
```suggestion
        :return: ``True`` if the MLflow client is running in this context. ``False`` otherwise.
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:21]:
```suggestion
        Determine if the MLflow client is running in this context.
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:30]:
```suggestion
        Provide the MLflow Experiment ID for the current context. Assumes that ``in_context()`` is ``True``.
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:32]:
```suggestion
        :return: The ID of the MLflow Experiment associated with the current context.
```

File: mlflow/tracking/_default_experiment/databricks_job_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/databricks_job_context.py:15]:
```suggestion
class DatabricksJobExperimentProvider(DefaultExperimentProvider):
```

File: mlflow/tracking/_default_experiment/databricks_job_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/databricks_job_context.py:35]:
```suggestion
        _logger.debug(
            "Job experiment with experiment ID '%s' fetched or created",
            experiment_id,
        )
```

File: mlflow/tracking/_default_experiment/databricks_job_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/databricks_job_context.py:25]: Can we add inline comments to https://github.com/mlflow/mlflow/blob/3c434246f73b41835194f3e6aa1f9be8488bcd56/mlflow/utils/mlflow_tags.py#L44-L45 explaining the content difference between these two tags?

File: mlflow/tracking/_default_experiment/databricks_notebook_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/databricks_notebook_context.py:5]:
```suggestion
class DatabricksNotebookExperimentProvider(DefaultExperimentProvider):
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [reply mlflow/tracking/_default_experiment/abstract_context.py:13]: Can we also explain that this is used to resolve a default experiment when the user has not explicitly set one themselves?

Comment (dbczumar) [reply mlflow/tracking/_default_experiment/abstract_context.py:13]: Can we lengthen the line width here?

File: mlflow/tracking/_default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/registry.py:23]: Can we increase the line width here?

File: mlflow/tracking/_default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/registry.py:21]:
```suggestion
    This class allows the registration of default experiment providers,
    which are used to provide MLflow Experiment IDs based on the current
    context where the MLflow client is running when the user has not
    explicitly set an experiment. Implementations declared though the
```

File: mlflow/tracking/_default_experiment/registry.py
Comment (dbczumar) [reply mlflow/tracking/_default_experiment/registry.py:64]: I think we can just remove this TODO. "0" is the default experiment ID expected / used by MLflow clients when no other context-specific information is available. We likely don't want to change this behavior until at least MLflow 2.0 for backwards compatibility reasons.

File: mlflow/tracking/_default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/registry.py:66]: Following from https://github.com/mlflow/mlflow/pull/5309/files#r793101898, I think we can rename this variable:

```suggestion
    experiment_id = "0"
```

File: mlflow/tracking/_default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/registry.py:70]: If a provider successfully returns an experiment ID, we should break out of the loop and return. Can we change this behavior and add a test case ensuring that the first successful provider wins?

File: tests/tracking/_default_experiment/test_databricks_job_context.py
Comment (dbczumar) [inline tests/tracking/_default_experiment/test_databricks_job_context.py:1]: Can we rename this file to `test_databricks_job_experiment_provider`?

File: mlflow/tracking/_default_experiment/databricks_job_context.py
Comment (dbczumar) [reply mlflow/tracking/_default_experiment/databricks_job_context.py:15]: Can we rename the file too? `databricks_job_experiment_provider.py`?

File: mlflow/tracking/_default_experiment/databricks_notebook_context.py
Comment (dbczumar) [reply mlflow/tracking/_default_experiment/databricks_notebook_context.py:5]: Can we rename the file too? `databricks_notebook_experiment_provider.py`?

File: tests/tracking/_default_experiment/test_databricks_notebook_context.py
Comment (dbczumar) [inline tests/tracking/_default_experiment/test_databricks_notebook_context.py:1]: Can we rename this file to `test_databricks_notebook_experiment_provider.py`?

File: tests/tracking/_default_experiment/test_databricks_notebook_context.py
Comment (dbczumar) [inline tests/tracking/_default_experiment/test_databricks_notebook_context.py:10]: Can we add a negative test case asserting that `in_context()` returns `False` by default and when `is_in_databricks_notebook()` returns `False`?

File: tests/tracking/_default_experiment/test_databricks_job_context.py
Comment (dbczumar) [inline tests/tracking/_default_experiment/test_databricks_job_context.py:24]: Can we add a negative test case asserting that `in_context()` returns `False` by default and when `is_in_databricks_job` returns `False` and when `get_job_type_info` returns something other than `NORMAL`?

File: tests/tracking/fluent/test_fluent.py
Comment (dbczumar) [inline tests/tracking/fluent/test_fluent.py:267]: Can we remove the print statements from this line and lines 269 / 280?

File: tests/tracking/fluent/test_fluent.py
Comment (dbczumar) [inline tests/tracking/fluent/test_fluent.py:186]: Why are we removing this test case? Seems useful.

File: tests/tracking/fluent/test_fluent.py
Comment (dbczumar) [inline tests/tracking/fluent/test_fluent.py:250]: Can we add a test case verifying that `get_experiment_by_id` delegates to `DefaultExperimentProvider` as expected? This is necessary to replace the coverage that we're losing by removing this test case.

**Review by apurva-koti — COMMENTED** [review]
General question, is it possible with this approach to resolve experiment IDs in a specific precedence order? For instance, we always want to go with a Jobs experiment first, even if we're in a notebook or repo notebook

**Comment by jinzhang21** [comment]
We probably want to do always test if the run context is automated (job). If not, then test if it's a repo notebook.

**Comment by apurva-koti** [comment]
@jinzhang21 yeah, in decreasing order of precedence it will be job -> repo notebook -> notebook. Job being first was briefly discussed in the review

File: tests/tracking/_default_experiment/test_databricks_notebook_context.py
Comment (sunishsheth2009) [reply tests/tracking/_default_experiment/test_databricks_notebook_context.py:10]: It should work since it would return whatever the value that is returned by `in_notebook_mock`. So it could either be true or false. both tests case should be tested as far as I can tell. Let me know if I am missing something

**Comment by sunishsheth2009** [comment]
> @jinzhang21 yeah, in decreasing order of precedence it will be job -> repo notebook -> notebook. Job being first was briefly discussed in the review

I think I added a test that the once we have a context that returns true, we get the experiment_id and we break the loop to check in other providers. So I think that should solve this problem as far as I could tell

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (sunishsheth2009) [reply mlflow/tracking/_default_experiment/abstract_context.py:4]: Nope. Sorry was on a break. But I have updated the PR with resolving comments. Hoping to push this soon. :)

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (jinzhang21) [inline mlflow/tracking/_default_experiment/abstract_context.py:1]: On filename, what's the reason to add "_" to "default_experiment"? This could be useful concept not just for Databricks. At least per @prakharg-msft 's comment it'd be useful for MSFT as well.

**Review by jinzhang21 — COMMENTED** [review]
Overall looks good on first scan. I'll take another detailed look later. Thanks for getting this started, Sunish!

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:15]:
```suggestion
    Usually the experiment_id is set explicitly by the user, but, if the experiment is not set,
    MLflow computes a default experiment id based on different contexts.
    When an experiment is created via the fluent ``mlflow.start_run`` method, MLflow iterates
    through the registered ``DefaultExperimentProvider``s until it finds one whose ``in_context()``
    method returns ``True``; MLflow then calls the provider's ``get_experiment_id()`` method and
    uses the resulting experiment ID for Tracking operations.
```

File: mlflow/tracking/_default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/abstract_context.py:35]: Can we fix the indentation here?

File: mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py
Comment (dbczumar) [inline mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py:28]: Can we add an inline comment that this is a "get or create" operation on Databricks when the `JOB` source type is specified? This will help future readers of the code understand what's happening.

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:1337]: For completeness, can we introduce an `EnvironmentVariableExperimentProvider`?

File: mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py
Comment (apurva-koti) [reply mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py:28]: No strong preference here, but isn't "get or create" not the client's concern? The client is just asking the backend to create an experiment and give it an ID to work with; details of where exactly that ID came from are abstracted away.

File: mlflow/tracking/fluent.py
Comment (dbczumar) [inline mlflow/tracking/fluent.py:1338]: `default_experiment_registry.get_experiment_id()` may be a resource-intensive or long-running operation. For example, the `DatabricksJobExperimentProvider` makes an API request every time this method is called. 

Can we set `_active_experiment_id` to the result of `default_experiment_registry.get_experiment_id()` to avoid, e.g., making a `CreateExperiment` API call every time `_get_experiment_id()` is invoked?

File: mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py
Comment (dbczumar) [reply mlflow/tracking/_default_experiment/databricks_job_experiment_provider.py:28]: It's our concern as maintainer's of the Databricks Jobs + Experiments integration, since we've gone ahead and implemented undocumented backend API behavior that a reader of this client code wouldn't otherwise be able to understand without sifting through the implementation of the Databricks MLflow Tracking service. All I'm asking for is some clarifying context.

File: mlflow/tracking/fluent.py
Comment (sunishsheth2009) [reply mlflow/tracking/fluent.py:1337]: I don't see the in_context pattern here to use the provider pattern for this. Let me know if I am missing something.

Comment (dbczumar) [reply mlflow/tracking/fluent.py:1337]: `"MLFLOW_EXPERIMENT_NAME" in os.environ`

File: mlflow/tracking/fluent.py
Comment (dbczumar) [reply mlflow/tracking/fluent.py:1338]: Edit: It's important to let users make changes to environment variable configs and have them be reflected. Instead, of my suggestion above, can we cache within `DatabricksJobExperimentProvider`?

File: mlflow/tracking/fluent.py
Comment (sunishsheth2009) [reply mlflow/tracking/fluent.py:1337]: Discussed offline that we can't refactor this code based on how the code is currently implemented

File: mlflow/tracking/default_experiment/abstract_context.py
Comment (dbczumar) [inline mlflow/tracking/default_experiment/abstract_context.py:27]: Nit: Can we fix the indentation here? `can` should be under ``True``.

File: mlflow/tracking/default_experiment/databricks_job_experiment_provider.py
Comment (dbczumar) [inline mlflow/tracking/default_experiment/databricks_job_experiment_provider.py:34]: Can we update this to more clearly explain that the behavior of `CreateExperiment` on Databricks is "get or create"? Right now, this comment implies that the client should be doing something different / special in each case, which makes the current implementation look like a bug.

File: mlflow/tracking/default_experiment/databricks_job_experiment_provider.py
Comment (dbczumar) [inline mlflow/tracking/default_experiment/databricks_job_experiment_provider.py:12]: We don't necessarily know if it's active or not (the user might have set a different experiment manually).

```suggestion
_resolved_job_experiment_id = None
```

File: mlflow/tracking/default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/default_experiment/registry.py:61]: "for the run" is a bit limiting. This routine may be used elsewhere. Can we remove references to "run context"?

File: mlflow/tracking/default_experiment/registry.py
Comment (dbczumar) [inline mlflow/tracking/default_experiment/registry.py:57]:
```suggestion
    """
    Get an experiment ID for the current context. The experiment ID is fetched by querying providers
    in the order that they were registered.
```

**Review by dbczumar — APPROVED** [review]
LGTM with some minor docstring nits. Thanks @sunishsheth2009 ! Can we make sure to test this manually before merging?

File: mlflow/tracking/fluent.py
Comment (jinzhang21) [inline mlflow/tracking/fluent.py:1346]: We have a TODO here to replace this with None. I think we should address it here instead of kicking the can down the road.

File: tests/tracking/default_experiment/test_registry.py
Comment (jinzhang21) [inline tests/tracking/default_experiment/test_registry.py:43]: Can we also add a test on testing the ordering of provider registration is honored correctly? That means when both `DatabricksNotebookExperimentProvider` and `DatabricksJobExperimentProvider` are in context, `DatabricksJobExperimentProvider` is returned.

**Review by jinzhang21 — COMMENTED** [review]
LGTM. Will approve after comments are addressed. Thanks, Sunish!

File: mlflow/tracking/fluent.py
Comment (sunishsheth2009) [reply mlflow/tracking/fluent.py:1346]: https://github.com/mlflow/mlflow/pull/5309#discussion_r793101898
Corey suggested that this would be a backwards incompat change and we would need to do this in 2.0

File: tests/tracking/default_experiment/test_registry.py
Comment (sunishsheth2009) [reply tests/tracking/default_experiment/test_registry.py:43]: There should already be a test for that here: https://github.com/mlflow/mlflow/pull/5309/files#diff-33f89924b0ce3f56c6d3e332907629a440d69b39e3c8e9c8dd3aff073d2b7d20R144
Let me know if I am missing something

File: mlflow/tracking/fluent.py
Comment (jinzhang21) [reply mlflow/tracking/fluent.py:1346]: SGTM. I guess we missed the 1.0 train.

**Review by jinzhang21 — APPROVED** [review]

**Pull request merged**
